// 3d-tiles-tools/build/traversal/ExplicitTraversedTile.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExplicitTraversedTile = void 0;
const ExplicitTraversedTiles_1 = require("./ExplicitTraversedTiles");
const MetadataSemanticOverrides_1 = require("./MetadataSemanticOverrides");
const ImplicitTilings_1 = require("../implicitTiling/ImplicitTilings");
/**
 * An implementation of a `TraversedTile` that reflects a tile
 * that actually appears as a JSON representation in the tileset.
 *
 * @internal
 */
class ExplicitTraversedTile {
    /**
     * Convenience function to create the root tile for a tile
     * traversal.
     *
     * @param root - The root tile from the tileset
     * @param schema - The optional metadata schema
     * @param resourceResolver - The `ResourceResolver` for
     * external references (like subtree files)
     * @returns The root `TraversedTile`
     */
    static createRoot(root, schema, resourceResolver) {
        const traversedRoot = new ExplicitTraversedTile(root, "/root", 0, undefined, schema, resourceResolver);
        return traversedRoot;
    }
    /**
     * Creates a new instance
     *
     * @param tile - The `Tile` from the tileset JSON
     * @param path - A JSON-path-like string describing this tile
     * @param level - The level, referring to the root of the
     * traversal, starting at 0
     * @param parent - The optional parent tile
     * @param schema - The optional metadata schema
     * @param resourceResolver - The `ResourceResolver` for
     * external references (like subtree files)
     */
    constructor(tile, path, level, parent, schema, resourceResolver) {
        // TODO For debugging
        this.toString = () => {
            return `ExplicitTraversedTile, level ${this.level}, path ${this.path}`;
        };
        this._tile = tile;
        this._path = path;
        this._level = level;
        this._parent = parent;
        this._schema = schema;
        this._resourceResolver = resourceResolver;
    }
    /**
     * Returns the `metadata` from the input JSON that defines the
     * `MetadataEntity` that is associated with this tile, or
     * `undefined` if the input did not contain a metadata entity.
     *
     * @returns The `MetadataEntity` object, or `undefined`
     */
    getMetadata() {
        return this._tile.metadata;
    }
    /**
     * Returns the `implicitTiling` from the input JSON that defines the
     * `TileImplicitTiling` that is associated with this tile, or
     * `undefined` if this tile does not define an implicit tiling.
     *
     * @returns The `TileImplicitTiling` object
     */
    getImplicitTiling() {
        return this._tile.implicitTiling;
    }
    /** {@inheritDoc TraversedTile.asRawTile} */
    asRawTile() {
        return this._tile;
    }
    /** {@inheritDoc TraversedTile.asFinalTile} */
    asFinalTile() {
        const tile = this._tile;
        const contents = this.getFinalContents();
        const finalTile = {
            boundingVolume: tile.boundingVolume,
            viewerRequestVolume: tile.viewerRequestVolume,
            geometricError: tile.geometricError,
            refine: tile.refine,
            transform: tile.transform,
            content: undefined,
            contents: contents,
            children: tile.children,
            metadata: tile.metadata,
            implicitTiling: tile.implicitTiling,
            extensions: tile.extensions,
            extras: tile.extras,
        };
        const schema = this._schema;
        if (schema) {
            MetadataSemanticOverrides_1.MetadataSemanticOverrides.applyExplicitTileMetadataSemanticOverrides(finalTile, schema);
        }
        return finalTile;
    }
    /** {@inheritDoc TraversedTile.path} */
    get path() {
        return this._path;
    }
    /** {@inheritDoc TraversedTile.level} */
    get level() {
        return this._level;
    }
    /** {@inheritDoc TraversedTile.getParent} */
    getParent() {
        return this._parent;
    }
    /** {@inheritDoc TraversedTile.getChildren} */
    async getChildren() {
        const implicitTiling = this._tile.implicitTiling;
        const schema = this._schema;
        if (implicitTiling) {
            const children = await ExplicitTraversedTiles_1.ExplicitTraversedTiles.createTraversedChildren(implicitTiling, schema, this, this._resourceResolver);
            return children;
        }
        if (!this._tile.children) {
            return [];
        }
        const children = this._tile.children;
        const childLevel = this._level + 1;
        const traversedChildren = [];
        for (let i = 0; i < children.length; i++) {
            const child = children[i];
            const childPath = this.path + "/children/" + i;
            const traversedChild = new ExplicitTraversedTile(child, childPath, childLevel, this, this._schema, this._resourceResolver);
            traversedChildren.push(traversedChild);
        }
        return traversedChildren;
    }
    /** {@inheritDoc TraversedTile.getRawContents} */
    getRawContents() {
        if (this._tile.content) {
            return [this._tile.content];
        }
        if (this._tile.contents) {
            return this._tile.contents;
        }
        return [];
    }
    /** {@inheritDoc TraversedTile.getFinalContents} */
    getFinalContents() {
        const rawContents = this.getRawContents();
        const schema = this._schema;
        if (!schema) {
            return rawContents;
        }
        const finalContents = [];
        for (let i = 0; i < rawContents.length; i++) {
            const rawContent = rawContents[i];
            const finalContent = {
                boundingVolume: rawContent.boundingVolume,
                uri: rawContent.uri,
                metadata: rawContent.metadata,
                group: rawContent.group,
                extensions: rawContent.extensions,
                extras: rawContent.extras,
            };
            MetadataSemanticOverrides_1.MetadataSemanticOverrides.applyExplicitContentMetadataSemanticOverrides(finalContent, schema);
            finalContents.push(finalContent);
        }
        return finalContents;
    }
    /** {@inheritDoc TraversedTile.getResourceResolver} */
    getResourceResolver() {
        return this._resourceResolver;
    }
    /** {@inheritDoc TraversedTile.isImplicitTilesetRoot} */
    isImplicitTilesetRoot() {
        return this._tile.implicitTiling !== undefined;
    }
    /** {@inheritDoc TraversedTile.getSubtreeUri} */
    getSubtreeUri() {
        const implicitTiling = this._tile.implicitTiling;
        if (!implicitTiling) {
            return undefined;
        }
        const rootCoordinates = ImplicitTilings_1.ImplicitTilings.createRootCoordinates(implicitTiling);
        const subtreeUri = ImplicitTilings_1.ImplicitTilings.substituteTemplateUri(implicitTiling.subdivisionScheme, implicitTiling.subtrees.uri, rootCoordinates);
        return subtreeUri;
    }
}
exports.ExplicitTraversedTile = ExplicitTraversedTile;
