// 3d-tiles-tools/build/spatial/QuadtreeCoordinates.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.QuadtreeCoordinates = void 0;
const MortonOrder_1 = require("./MortonOrder");
const Quadtrees_1 = require("./Quadtrees");
/**
 * An implementation of `TreeCoordinates` for octrees
 *
 * @internal
 */
class QuadtreeCoordinates {
    constructor(level, x, y) {
        this.toString = () => {
            return `(level ${this.level}, (${this._x},${this._y}))`;
        };
        this._level = level;
        this._x = x;
        this._y = y;
    }
    /** {@inheritDoc TreeCoordinates.level} */
    get level() {
        return this._level;
    }
    get x() {
        return this._x;
    }
    get y() {
        return this._y;
    }
    /** {@inheritDoc TreeCoordinates.parent} */
    parent() {
        if (this._level === 0) {
            return null;
        }
        const pLevel = this._level - 1;
        const px = this._x >> 1;
        const py = this._y >> 1;
        return new QuadtreeCoordinates(pLevel, px, py);
    }
    /** {@inheritDoc TreeCoordinates.children} */
    *children() {
        const nLevel = this._level + 1;
        const nX = this._x << 1;
        const nY = this._y << 1;
        yield new QuadtreeCoordinates(nLevel, nX + 0, nY + 0);
        yield new QuadtreeCoordinates(nLevel, nX + 1, nY + 0);
        yield new QuadtreeCoordinates(nLevel, nX + 0, nY + 1);
        yield new QuadtreeCoordinates(nLevel, nX + 1, nY + 1);
    }
    /** {@inheritDoc TreeCoordinates.descendants} */
    descendants(maxLevelInclusive, depthFirst) {
        const queue = [this];
        const result = {
            [Symbol.iterator]() {
                return this;
            },
            next() {
                const element = depthFirst ? queue.pop() : queue.shift();
                if (!element) {
                    return { value: undefined, done: true };
                }
                if (element.level < maxLevelInclusive) {
                    for (const c of element.children()) {
                        queue.push(c);
                    }
                }
                return { value: element, done: false };
            },
        };
        return result;
    }
    /** {@inheritDoc TreeCoordinates.toArray} */
    toArray() {
        return [this.level, this.x, this.y];
    }
    /** {@inheritDoc TreeCoordinates.toIndex} */
    toIndex() {
        const offset = Quadtrees_1.Quadtrees.computeNumberOfNodesForLevels(this._level);
        return offset + this.toIndexInLevel();
    }
    /** {@inheritDoc TreeCoordinates.toIndexInLevel} */
    toIndexInLevel() {
        return MortonOrder_1.MortonOrder.encode2D(this._x, this._y);
    }
}
exports.QuadtreeCoordinates = QuadtreeCoordinates;
