// 3d-tiles-tools/build/tilesetData/TilesetInMemory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TilesetInMemory = void 0;
const TilesetError_1 = require("./TilesetError");
/**
 * Implementation of a TilesetSource and TilesetTarget that
 * stores the data in memory.
 *
 * This is mainly intended for tests and debugging.
 *
 * @internal
 */
class TilesetInMemory {
    /**
     * Default constructor
     */
    constructor() {
        /**
         * The mapping from keys to the actual data
         */
        this.dataMap = {};
        this.sourceIsOpen = false;
        this.targetIsOpen = false;
        this.overwrite = false;
    }
    /** {@inheritDoc TilesetSource.open} */
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    open(fullInputName) {
        if (this.sourceIsOpen) {
            throw new TilesetError_1.TilesetError("Source already opened");
        }
        this.sourceIsOpen = true;
    }
    /** {@inheritDoc TilesetSource.getKeys} */
    getKeys() {
        if (!this.sourceIsOpen) {
            throw new TilesetError_1.TilesetError("Source is not opened. Call 'open' first.");
        }
        return Object.keys(this.dataMap).values();
    }
    /** {@inheritDoc TilesetSource.getValue} */
    getValue(key) {
        if (!this.sourceIsOpen) {
            throw new TilesetError_1.TilesetError("Source is not opened. Call 'open' first.");
        }
        return this.dataMap[key];
    }
    /** {@inheritDoc TilesetSource.close} */
    close() {
        if (!this.sourceIsOpen) {
            throw new TilesetError_1.TilesetError("Source is not opened. Call 'open' first.");
        }
        this.sourceIsOpen = false;
    }
    /** {@inheritDoc TilesetTarget.begin} */
    begin(fullOutputName, overwrite) {
        if (this.targetIsOpen) {
            throw new TilesetError_1.TilesetError("Target already opened");
        }
        this.targetIsOpen = true;
        this.overwrite = overwrite;
    }
    /** {@inheritDoc TilesetTarget.addEntry} */
    addEntry(key, content) {
        if (!this.targetIsOpen) {
            throw new TilesetError_1.TilesetError("Target is not opened. Call 'begin' first.");
        }
        if (this.dataMap[key]) {
            if (!this.overwrite) {
                throw new TilesetError_1.TilesetError("Entry already exists: " + key);
            }
        }
        this.dataMap[key] = content;
    }
    /** {@inheritDoc TilesetTarget.end} */
    async end() {
        if (!this.targetIsOpen) {
            throw new TilesetError_1.TilesetError("Target is not opened. Call 'begin' first.");
        }
        this.targetIsOpen = false;
        this.overwrite = false;
    }
}
exports.TilesetInMemory = TilesetInMemory;
