// 3d-tiles-tools/build/tilesetData/TilesetTargetFs.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TilesetTargetFs = void 0;
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
const TilesetError_1 = require("./TilesetError");
/**
 * Implementation of a TilesetTarget that writes into
 * a directory of a file system
 *
 * @internal
 */
class TilesetTargetFs {
    /**
     * Default constructor
     */
    constructor() {
        this.fullOutputName = undefined;
        this.overwrite = false;
    }
    /** {@inheritDoc TilesetTarget.begin} */
    begin(fullOutputName, overwrite) {
        if (this.fullOutputName) {
            throw new TilesetError_1.TilesetError("Target already opened");
        }
        this.fullOutputName = fullOutputName;
        this.overwrite = overwrite;
        if (!fs_1.default.existsSync(fullOutputName)) {
            fs_1.default.mkdirSync(fullOutputName, { recursive: true });
        }
    }
    /** {@inheritDoc TilesetTarget.addEntry} */
    addEntry(key, content) {
        if (!this.fullOutputName) {
            throw new TilesetError_1.TilesetError("Target is not opened. Call 'begin' first.");
        }
        const fullOutputFileName = path_1.default.join(this.fullOutputName, key);
        if (fs_1.default.existsSync(fullOutputFileName)) {
            if (!this.overwrite) {
                throw new TilesetError_1.TilesetError("File already exists: " + fullOutputFileName);
            }
        }
        // TODO Need to unlink if file exists?
        fs_1.default.mkdirSync(path_1.default.dirname(fullOutputFileName), { recursive: true });
        fs_1.default.writeFileSync(fullOutputFileName, content);
    }
    /** {@inheritDoc TilesetTarget.end} */
    async end() {
        if (!this.fullOutputName) {
            throw new TilesetError_1.TilesetError("Target is not opened. Call 'begin' first.");
        }
        this.fullOutputName = undefined;
        this.overwrite = false;
    }
}
exports.TilesetTargetFs = TilesetTargetFs;
